client-sdk/ts-web/core/src/hdkey.ts

```diff
@@ -0,0 +1,93 @@
+import {sha512} from 'js-sha512';
+import {SignKeyPair, sign} from 'tweetnacl';
+import {generateMnemonic, mnemonicToSeed} from 'bip39';
+
+const ED25519_CURVE = 'ed25519 seed';
+const HARDENED_OFFSET = 0x80000000;
+const replaceDerive = (val: string): string => val.replace("'", '');
+const pathRegex = new RegExp("^m(\\/[0-9]+')+$");
+
+export class HDKey {
+    public readonly keypair: SignKeyPair;
+
+    constructor(public readonly privateKey: Uint8Array, public readonly chainCode: Uint8Array) {
+        this.keypair = sign.keyPair.fromSeed(privateKey);
+    }
+
+    /**
+     * Returns an instance of HDKey that can then be used for derivation
+     * @param seed Master seed for key derivation
+     * @returns Instance of HDKey
+     */
+    public static fromSeed(seed: Uint8Array): HDKey {
+        const hash = sha512.hmac.arrayBuffer(ED25519_CURVE, Buffer.from(seed));
+
+        const I = new Uint8Array(hash);
+        const IL = I.slice(0, 32);
+        const IR = I.slice(32);
+
+        return new HDKey(IL, IR);
+    }
+
+    /**
+     * Returns the root node for the given mnemonic
+     * @param mnemonic Bip39 Mnemonic
+     * @param passphrase Optional passphrase
+     * @returns Root node
+     */
+    public static async fromMnemonic(mnemonic: string, passphrase?: string): Promise<HDKey> {
+        const seed = await mnemonicToSeed(mnemonic, passphrase);
+        return HDKey.fromSeed(seed);
+    }
+
+    /**
+     * Generates a mnemonic
+     * @param strength Length in bits of the generated mnemonic
+     * @returns Generated BIP39 Mnemonic
+     */
+    public static generateMnemonic(strength = 256): string {
+        return generateMnemonic(strength);
+    }
+
+    /**
+     * Returns the HDKey for the given derivation path
+     * @param path Derivation path, starting with m/
+     * @returns Instance of HDKey
+     */
+    public derivePath(path: string): HDKey {
+        if (!path.match(pathRegex)) {
+            throw new Error(
+                "Invalid derivation path. Valid paths must use a format similar to : m/44'/474'/0' and all indexes must be hardened",
+            );
+        }
+
+        const segments = path
+            .split('/')
+            .slice(1)
+            .map(replaceDerive)
+            .map((el) => parseInt(el, 10));
+
+        return segments.reduce((parent, segment) => parent.derive(segment + HARDENED_OFFSET), this);
+    }
+
+    /**
+     * Derive the child key at the specified index
+     * @param index
+     * @returns Instance of HDKey
+     */
+    protected derive(index: number): HDKey {
+        const indexBuffer = Buffer.allocUnsafe(4);
+        indexBuffer.writeUInt32BE(index, 0);
+
+        const data = Buffer.concat([Buffer.alloc(1, 0), this.privateKey, indexBuffer]);
```

Comment:
If you care enough about performance to `allocUnsafe`, why not pre-alloc and write the index (and prefix) at the correct offset?

Reply:
Do you mean this ?

```typescript
    const data = Buffer.allocUnsafe(37);
    data.writeUInt8(0, 0)
    data.fill(this.privateKey, 1, 33)
    data.writeUInt32BE(index, 33);
```

Reply:
yeah that looks about right

Reply:
After careful consideration and benchmarking I think it's better to keep it as it is as it's more performant while also being easier to read

```javascript
const Benchmark = require('benchmark');
const suite = new Benchmark.Suite;

const index = 0;
const pk = new Uint8Array(32)

suite
.add('Buffer#prealloc', function() {
    const data = Buffer.allocUnsafe(37);
    data.writeUInt8(0, 0);
    data.fill(pk, 1, 33)
    data.writeUInt32BE(index, 33);
})
.add('Buffer#concat', function() {
    const indexBuffer = Buffer.allocUnsafe(4);
    indexBuffer.writeUInt32BE(index, 0);

    const data = Buffer.concat([Buffer.alloc(1, 0), pk, indexBuffer]);
})
/*
Buffer#prealloc x 3,620,673 ops/sec ±0.34% (292 runs sampled)
Buffer#concat x 4,385,810 ops/sec ±0.35% (291 runs sampled)
Fastest is Buffer#concat
*/
````